group_theory/group_action.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import data.set.finite group_theory.coset
+
+universes u v
+variables {α : Type u} {β : Type v}
+
+class is_monoid_action [monoid α] (f : α → β → β) : Prop :=
+(one : ∀ a : β, f (1 : α) a = a)
+(mul : ∀ (x y : α) (a : β), f (x * y) a = f x (f y a))
+
+namespace is_monoid_action
+
+variables [monoid α] (f : α → β → β) [is_monoid_action f] 
+
+def orbit (a : β) := set.range (λ x : α, f x a)
+
+lemma mem_orbit_iff {f : α → β → β} [is_monoid_action f] {a b : β} :
+  b ∈ orbit f a ↔ ∃ x : α, f x a = b :=
+iff.rfl
+
+@[simp] lemma mem_orbit (a : β) (x : α) :
+  f x a ∈ orbit f a :=
+⟨x, rfl⟩
+
+@[simp] lemma mem_orbit_self (a : β) :
+  a ∈ orbit f a :=
+⟨1, show f 1 a = a, by simp [is_monoid_action.one f]⟩
+
+instance orbit_fintype (a : β) [fintype α] [decidable_eq β] :
+  fintype (orbit f a) := set.fintype_range _
+
+def stabilizer (a : β) : set α :=
+{x : α | f x a = a}
+
+lemma mem_stabilizer_iff {f : α → β → β} [is_monoid_action f] {a : β} {x : α} :
+  x ∈ stabilizer f a ↔ f x a = a :=
+iff.rfl
+
+def fixed_points : set β := {a : β | ∀ x, x ∈ stabilizer f a}
+
+lemma mem_fixed_points {f : α → β → β} [is_monoid_action f] {a : β} :
+  a ∈ fixed_points f ↔ ∀ x : α, f x a = a := iff.rfl
+
+lemma mem_fixed_points' {f : α → β → β} [is_monoid_action f] {a : β} : a ∈ fixed_points f ↔
+  (∀ b, b ∈ orbit f a → b = a) :=
+⟨λ h b h₁, let ⟨x, hx⟩ := mem_orbit_iff.1 h₁ in hx ▸ h x,
+λ h b, mem_stabilizer_iff.2 (h _ (mem_orbit _ _ _))⟩
+
+end is_monoid_action
+
+namespace is_group_action
+variables [group α] (f : α → β → β)
+open is_monoid_action
+
+class is_group_action extends is_monoid_action f
+
+variable [is_group_action f]
+
+def to_perm (g : α) : equiv.perm β :=
+{ to_fun := f g,
+  inv_fun := f g⁻¹,
+  left_inv := λ a, by rw [← is_monoid_action.mul f, inv_mul_self, is_monoid_action.one f],
+  right_inv := λ a, by rw [← is_monoid_action.mul f, mul_inv_self, is_monoid_action.one f] }
+
+instance : is_group_hom (to_perm f) :=
+{ mul := λ x y, equiv.ext _ _ (λ a, is_monoid_action.mul f x y a) }
+
+lemma bijective (g : α) : function.bijective (f g) :=
+(to_perm f g).bijective
+
+lemma orbit_eq_iff {f : α → β → β} [is_monoid_action f] {a b : β} : 
+  a ∈ orbit f b ↔ orbit f a = orbit f b :=
```

Comment:
If you want to call this `orbit_eq_iff`, the equality should be on the left. In this direction it would be more like `mem_orbits_iff_eq` or a variant.